backend/test_observer/controllers/test_results/test_results.py

```diff
@@ -0,0 +1,307 @@
+#
+# This file is part of Test Observer Backend.
+#
+# Test Observer Backend is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# Test Observer Backend is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from datetime import datetime
+
+from fastapi import APIRouter, Depends, Query
+from sqlalchemy import and_, desc, distinct, select, func
+from sqlalchemy.orm import Session, joinedload
+
+from test_observer.data_access.models import (
+    Artefact,
+    ArtefactBuild,
+    Environment,
+    TestExecution,
+    TestResult,
+    TestCase,
+    Issue,
+    IssueTestResultAttachment,
+)
+from test_observer.data_access.models_enums import FamilyName
+from test_observer.data_access.setup import get_db
+
+from .models import TestResultSearchResponse
+
+router = APIRouter(tags=["test-results"])
+
+
+@router.get("", response_model=TestResultSearchResponse)
+def search_test_results(
+    family: str | None = Query(
+        None, description="Filter by artefact family (e.g., Charm, Snap)"
+    ),
+    environment: str | None = Query(
+        None,
+        description="Filter by environment name (e.g., Juju:3/stable ubuntu:20.04)",
+    ),
+    test_case: str | None = Query(
+        None, description="Filter by test case name (e.g., test_deploy)"
+    ),
+    template_id: str | None = Query(None, description="Filter by template ID"),
+    issues: str | None = Query(
+        None, description="Filter by Jira or GitHub issue IDs (comma-separated)"
+    ),
+    from_date: datetime | None = Query(
+        None, description="Filter results from this timestamp"
+    ),
+    until: datetime | None = Query(
+        None, description="Filter results until this timestamp"
+    ),
+    limit: int = Query(
+        50, ge=1, le=1000, description="Maximum number of results to return"
+    ),
+    offset: int = Query(
+        0, ge=0, description="Number of results to skip for pagination"
+    ),
+    db: Session = Depends(get_db),
+) -> TestResultSearchResponse:
+    """
+    Search test results across artefacts using flexible filters.
+
+    This endpoint uses a single optimized query with window functions to get both
+    the total count and paginated results in one database round trip.
+    """
+
+    # Build filters
+    filters = []
+
+    if family:
+        try:
+            family_enum = FamilyName(family.lower())
+            filters.append(Artefact.family == family_enum)
+        except ValueError:
+            filters.append(Artefact.family.is_(None))
```

Comment:
I think this should return a bad request if the family is invalid. You should be able to use the enum directly above though, not need to parse it here.